gateway-service-impl/src/main/java/org/hypertrace/gateway/service/entity/query/ExecutionTreeUtils.java

```diff
@@ -179,4 +187,131 @@ private static boolean containsNotOrOrFilter(Filter filter) {
 
     return false;
   }
+
+  /**
+   * Computes common set of sources, if both filters and order bys are requested on the same source
+   * sets
+   *
+   * Look at {@link ExecutionTreeUtils#getIntersectingSourceSets(java.util.Map, java.util.Map)}
+   */
+  public static Set<String> getSourceSetsIfFilterAndOrderByAreFromSameSourceSets(
+      ExecutionContext executionContext) {
+    Map<String, List<Expression>> sourceToFilterExpressionMap =
+        executionContext.getSourceToFilterExpressionMap();
+    Map<String, List<OrderByExpression>> sourceToOrderByExpressionMap =
+        executionContext.getSourceToOrderByExpressionMap();
+
+    // A weird case, if there are no filters and order bys
+    if (sourceToFilterExpressionMap.isEmpty() && sourceToOrderByExpressionMap.isEmpty()) {
+      return Collections.emptySet();
+    }
+
+    Map<String, Set<String>> filterAttributeToSourcesMap =
+        buildAttributeToSourcesMap(sourceToFilterExpressionMap);
+    Map<String, Set<String>> orderByAttributeToSourceMap =
+        buildAttributeToSourcesMap(
+            sourceToOrderByExpressionMap.entrySet().stream()
+                .collect(
+                    Collectors.toMap(
+                        Map.Entry::getKey,
+                        entry ->
+                            entry.getValue().stream()
+                                .map(OrderByExpression::getExpression)
+                                .collect(Collectors.toList()))));
+
+    return getIntersectingSourceSets(filterAttributeToSourcesMap, orderByAttributeToSourceMap);
+  }
+
+  /**
+   * Given a source to expression map, builds an attribute to sources map, where attribute value is extracted
+   * out from the expression. Basically, a reverse map of the map provided as input
+   *
+   * Example:
+   * ("QS" -> API.id, "QS" -> API.name, "EDS" -> API.id) =>
+   *
+   * ("API.id" -> ["QS", "EDS"], "API.name" -> "QS")
+   */
+  private static Map<String, Set<String>> buildAttributeToSourcesMap(
```

Comment:
Isn't this map already there in ExecutionContext?
See attributeToSourcesMap in ExecutionContext

Reply:
Wondering why I missed that. But, yes, I think I can use that to filter out the filter and order by attributes, instead of building the entire map again